charms/istio-pilot/src/charm.py

```diff
@@ -130,54 +132,91 @@ def remove(self, event):
-                }
+                {"service-name": f'istiod.{self.model.name}.svc', "service-port": '15012'}
             )
 
     def handle_ingress(self, event):
+        try:
+            self._get_gateway_address
+        except (ApiError, TypeError) as e:
+            if e == ApiError:
+                self.log.exception("ApiError: Could not get istio-ingressgateway, retrying")
+            elif e == TypeError:
+                self.log.exception("TypeError: No ip address found, retrying")
+            event.defer()
+            return
+        else:
+            self.unit.status = ActiveStatus()
+
         ingress = self.interfaces['ingress']
 
         if ingress:
-            routes = ingress.get_data().items()
+            # Filter out data we sent back.
+            routes = {
+                (rel, app): route
+                for (rel, app), route in sorted(
+                    ingress.get_data().items(), key=lambda tup: tup[0][0].id
+                )
+                if app != self.app
+            }
         else:
-            routes = []
+            routes = {}
+
+        if isinstance(event, (RelationBrokenEvent)):
+            # The app-level data is still visible on a broken relation, but we
+            # shouldn't be keeping the VirtualService for that related app.
+            del routes[(event.relation, event.app)]
 
         t = self.env.get_template('virtual_service.yaml.j2')
-        gateway = self.model.config['default-gateways'].split(',')[0]
+        gateway = self.model.config['default-gateway']
 
         def get_kwargs(version, route):
             """Handles both v1 and v2 ingress relations.
 
             v1 ingress schema doesn't allow sending over a namespace.
             """
-            kwargs = {'gateway': gateway, **route}
+            kwargs = {'gateway': gateway, 'app_name': self.app.name, **route}
 
             if 'namespace' not in kwargs:
                 kwargs['namespace'] = self.model.name
 
             return kwargs
 
-        virtual_services = ''.join(
-            t.render(**get_kwargs(ingress.versions[app.name], route))
-            for ((_, app), route) in routes
+        virtual_services = '\n---'.join(
+            t.render(**get_kwargs(ingress.versions[app.name], route)).strip().strip("---")
+            for ((_, app), route) in routes.items()
         )
 
-        t = self.env.get_template('gateway.yaml.j2')
-        gateways = self.model.config['default-gateways'].split(',')
-        gateways = ''.join(t.render(name=g) for g in gateways)
-
-        manifests = [virtual_services, gateways]
-        manifests = '\n'.join([m for m in manifests if m])
-
-        for resource in [self.virtual_service_resource, self.gateway_resource]:
-            self._delete_existing_resource_objects(resource, namespace=self.model.name)
+        self._delete_existing_resource_objects(
+            self.virtual_service_resource, namespace=self.model.name
+        )
```

Comment:
Am I understanding correctly that we are deleting all the virtual services and recreating them each time we call `handle_ingress`? Is this necessary?

Reply:
Yes, we are deleting and creating VirtualServices every time. Regarding the need for it, I think @ca-scribner can elaborate more. I'm not sure if it is because we didn't have an `apply` option in lightkube. 
Also, there's an important concern raised [here](https://github.com/canonical/istio-operators/issues/50) regarding the implications of doing it this way, if we want to get rid of it, that could be a good reason.

Reply:
Yes that was my concern too, specially considering that some charms might be very noisy with this relation.

Reply:
Yeah I think this was more challenging before using lightkube and so Ken used this as a easy first implementation.

Lightkube's `apply` makes `add` and `update` actions really easy, but we also need to cover `delete`.  For an `ingress` relation that is dropped, our relation data bag just wont have that relation, so doing a simple `apply` of the metadata would update new/existing `VirtualServices` but would just leave `VirtualService`s for removed relations in place (since an `apply` never deletes anything).  But again with lightkube, this is pretty easy to cover - one solution would be:
* lightkube.get all VirtualServices **created by this charm** (we don't want to delete virtualservices from other stuff like seldon or knative)
* compare existing `VirtualService`s to desired, deleting any that exist now but are not backed by an existing relation
* apply the metadata for new/existing

Reply:
(Identifying which `vs`s are from this charm vs other sources was a problem in the past too and is addressed in this charm by adding the (poorly named - please rename if you think of something better) label `app.istio-pilot.io/is-workload-entity=true`)

Reply:
I'd like to explore the option of getting the `virtualservices` created by this charm, comparing desired against existing ones without a backing relation (delete those) and apply new/existing. That is addressing a different issue, though, I'd like to have a separate PR just for that.